node/hyperbahn/rate_limiting.js

```diff
@@ -0,0 +1,187 @@
+            counter.qps -= counter.buckets[next];
+        }
+        assert(counter.qps >= 0, 'qps shouls always be larger equal to 0');
+        counter.index = next;
+        counter.buckets[counter.index] = 0;
+        if (serviceName === TOTAL_QPS) {
+            // total QPS limit never change
+            continue;
+        }
+        var limit = self.qpsLimits[serviceName] || self.defaultServiceQpsLimit;
+        counter.qpsLimit = limit / self.egressNodes.kValueFor(serviceName);
+    }
+
+    self.refreshTimer = self.timers.setTimeout(
+        function refresh() {
+            self.refresh();
+        },
+        self.refreshDelay
+    );
+};
+
+RateLimiting.prototype.removeCounter =
+function removeCounter(serviceName) {
+    var self = this;
+    delete self.counters[serviceName];
+};
+
+RateLimiting.prototype.isExitFor =
+function isExitFor(serviceName) {
+    var self = this;
+
+    var chan = self.serviceDispatchHandler.channel.subChannels[serviceName];
+    if (!chan) {
+        return self.egressNodes.isExitFor(serviceName);
+    }
+
+    return chan.serviceProxyMode === 'exit';
+};
+
+RateLimiting.prototype.createCounter =
+function createCounter(serviceName) {
+    var self = this;
+
+    var counter = Object.create(null);
+    counter.index = 0;
+    counter.qps = 0;
+    counter.buckets = [];
+    counter.buckets.length = self.buckets;
+    counter.buckets[0] = 0;
+    if (serviceName === TOTAL_QPS) {
+        counter.qpsLimit = self.totalQpsLimit;
+    } else {
+        var limit = self.qpsLimits[serviceName] || self.defaultServiceQpsLimit;
+        counter.qpsLimit = limit / self.egressNodes.kValueFor(serviceName);
+    }
+
+    return counter;
+};
+
+RateLimiting.prototype.incrementCounter =
+function incrementCounter(serviceName) {
+    var self = this;
+
+    var counter = self.counters[serviceName];
+    if (!counter) {
+        counter = self.createCounter(serviceName);
+        self.counters[serviceName] = counter;
+    }
+
+    counter.buckets[counter.index] += 1;
+    counter.qps += 1;
+};
+
+RateLimiting.prototype.shouldRateLimit =
+function shouldRateLimit(serviceName) {
+    var self = this;
+
+    var counter = self.counters[serviceName];
+    assert(counter, 'cannot find counter for ' + serviceName);
+    return counter.qps <= counter.qpsLimit ? 0 : counter.qpsLimit;
+};
+
+RateLimiting.prototype.handleRequest =
+function handleRequest(req, buildRes) {
+    var self = this;
+    assert(req.serviceName, 'serviceName should not be empty for rate limiting.');
+
+    if (req.serviceName === 'hyperbahn') {
+        return false;
+    }
+
+    // check total QPS
+    self.incrementCounter(TOTAL_QPS);
+    var limit = self.shouldRateLimit(TOTAL_QPS);
+    if (limit) {
+        buildRes().sendError('Busy', 'hyperbahn node is rate limited by the total rate ' + limit + ' qps');
+        return true;
+    }
+
+    // check QPS for serviceName
+    if (!self.isExitFor(req.serviceName)) {
```

Comment:
Actually, I think rate limiting should be part of hyperbahn since tchannel is just a RPC call framework and should not have anything to do with rate limiting. @Raynos opinions?